File: 2022/day03.py
import fileinput

lines = [line.strip() for line in fileinput.input()]
silver = 0
for line in lines:
    s1, s2 = set(line[:len(line)//2]), set(line[len(line)//2:])
    r = s1.intersection(s2)
    assert (len(r) == 1)
    c = r.pop()
    o = ord(c)
    if ord('A') <= o <= ord('Z'):
        silver += 27 + o - ord('A')
    else:
        silver += 1 + o - ord('a')
gold = 0
for i in range(0, len(lines), 3):
    s1, s2, s3 = set(lines[i]), set(lines[i+1]), set(lines[i+2])
    r = s1.intersection(s2).intersection(s3)
    assert (len(r) == 1)
    c = r.pop()
    o = ord(c)
    if ord('A') <= o <= ord('Z'):
        gold += 27 + o - ord('A')
    else:
        gold += 1 + o - ord('a')
print(silver)
print(gold)

File: 2022/src/day03.rs
use std::collections::HashSet;

pub fn main() {
    let input = include_str!("../3");

    let s: u32 = input
        .lines()
        .map(|line| {
            let a: HashSet<char> = HashSet::from_iter(line.chars().take(line.len() / 2));
            let b: HashSet<char> = HashSet::from_iter(line.chars().skip(line.len() / 2));
            a.intersection(&b).next().unwrap().to_owned()
        })
        .map(|c| {
            let p = c as u32;
            p % 32 + 26 * (p >> 5 & 1 ^ 1) // thanks sheks
        })
        .sum();

    let g: u32 = input
        .lines()
        .step_by(3)
        .zip(input.lines().skip(1).step_by(3))
        .zip(input.lines().skip(2).step_by(3))
        .map(|((a, b), c)| {
            let mut a: HashSet<char> = HashSet::from_iter(a.chars());
            let b: HashSet<char> = HashSet::from_iter(b.chars());
            let c: HashSet<char> = HashSet::from_iter(c.chars());
            a.retain(|x| b.contains(x) && c.contains(x));
            a.iter().next().unwrap().to_owned()
        })
        .map(|c| {
            let p = c as u32;
            p % 32 + 26 * (p >> 5 & 1 ^ 1) // thanks sheks
        })
        .sum();

    println!("silver: {}", s);
    println!("gold: {}", g);
}
